ocaml/typing/typedecl.ml

```diff
@@ -2248,19 +2265,46 @@ let make_native_repr env core_type ty ~global_repr ~is_layout_poly ~why =
     *)
     | Tvar {jkind} when is_layout_poly
                       && Jkind.is_any jkind
-                      && get_level ty = Btype.generic_level -> Repr_poly
+                      && get_level ty = Btype.generic_level -> Poly
     | _ ->
       let sort =
         type_sort_external ~is_layout_poly ~why env core_type.ptyp_loc ty
       in
+      Sort sort
+  in
+  match get_native_repr_attribute
+          core_type.ptyp_attributes ~global_repr,
+        sort_or_poly with
+  | Native_repr_attr_absent, Poly ->
+    Repr_poly
+  | Native_repr_attr_absent, Sort (Value as sort) ->
+    Same_as_ocaml_repr sort
+  | Native_repr_attr_absent, (Sort sort) ->
+    if Language_extension.erasable_extensions_only ()
+    then
+      (* Non-value sorts without [@unboxed] are not erasable. *)
+      raise (Error (core_type.ptyp_loc,
+              Missing_unboxed_attribute_on_non_value_sort sort))
+    else
       Same_as_ocaml_repr sort
-    end
-  | Native_repr_attr_present kind ->
+  | Native_repr_attr_present kind, (Poly | Sort Value)
+  | Native_repr_attr_present (Untagged as kind), Sort _ ->
     begin match native_repr_of_type env kind ty with
     | None ->
       raise (Error (core_type.ptyp_loc, Cannot_unbox_or_untag_type kind))
     | Some repr -> repr
     end
+  | Native_repr_attr_present Unboxed, (Sort sort) ->
+    (* We allow [@unboxed] on non-value sorts. *)
```

Comment:
Please add a few sentences describing the design here. It's not really a perfect place to put that text, but someone reading this will be curious as to why this is, and writing it here will be helpful. Thanks.

Reply:
Added a short comment for why we want to allow `[@unboxed]` here